File: core/domain/src/main/kotlin/com/deepseek/harness/android/domain/model/Attachment.kt
package com.deepseek.harness.android.domain.model

/**
 * Image-attachment vocabulary for the version-one path. Raster images ride
 * the prompt wire as inline content parts; durable references come back in
 * the session log and download through `session.attachment`.
 */

/** One composer-held image awaiting send, encoded as base64. */
data class PendingImage(
    val id: String,
    val mediaType: String,
    val base64Data: String,
    val name: String? = null,
    val byteSize: Long = 0L,
)

/** Durable image reference found inside a message's content blocks. */
data class AttachmentRef(
    val attachmentId: String,
    val mediaType: String,
    val bytes: Long,
    val width: Int,
    val height: Int,
    val name: String? = null,
)

/** One downloaded image: its durable reference plus decoded bytes. */
class AttachmentData(
    val ref: AttachmentRef,
    val data: ByteArray,
) {
    override fun equals(other: Any?): Boolean =
        other is AttachmentData && other.ref == ref && other.data.contentEquals(data)

    override fun hashCode(): Int = ref.hashCode() * 31 + data.contentHashCode()
}

/**
 * Host admission limits for images, mirrored from the `imageLimits`
 * session projection. Null fields fall back to client-side defaults until
 * the host projection arrives.
 */
data class ImageLimits(
    val maxImageBytes: Long = DEFAULT_MAX_IMAGE_BYTES,
    val maxImagesPerMessage: Int = DEFAULT_MAX_IMAGES_PER_MESSAGE,
    val maxMessageImageBytes: Long = DEFAULT_MAX_IMAGE_BYTES,
    val maxImagePixels: Long = DEFAULT_MAX_IMAGE_PIXELS,
    val mediaTypes: List<String> = DEFAULT_MEDIA_TYPES,
) {
    companion object {
        const val DEFAULT_MAX_IMAGE_BYTES: Long = 5L * 1024 * 1024
        const val DEFAULT_MAX_IMAGES_PER_MESSAGE: Int = 20
        const val DEFAULT_MAX_IMAGE_PIXELS: Long = 30_000_000L
        val DEFAULT_MEDIA_TYPES: List<String> = listOf(
            "image/png",
            "image/jpeg",
            "image/webp",
            "image/gif",
        )
    }
}

File: flutter/packages/domain/lib/model/attachment.dart
/// Image-attachment vocabulary for the version-one path. Raster images ride
/// the prompt wire as inline content parts; durable references come back in
/// the session log and download through `session.attachment`.
library;

import 'dart:typed_data';

/// One composer-held image awaiting send, encoded as base64.
final class PendingImage {
  const PendingImage({
    required this.id,
    required this.mediaType,
    required this.base64Data,
    this.name,
    this.byteSize = 0,
  });

  final String id;
  final String mediaType;
  final String base64Data;
  final String? name;
  final int byteSize;

  @override
  bool operator ==(Object other) =>
      other is PendingImage &&
      other.id == id &&
      other.mediaType == mediaType &&
      other.base64Data == base64Data &&
      other.name == name &&
      other.byteSize == byteSize;

  @override
  int get hashCode => Object.hash(id, mediaType, base64Data, name, byteSize);
}

/// Durable image reference found inside a message's content blocks.
final class AttachmentRef {
  const AttachmentRef({
    required this.attachmentId,
    required this.mediaType,
    required this.bytes,
    required this.width,
    required this.height,
    this.name,
  });

  final String attachmentId;
  final String mediaType;
  final int bytes;
  final int width;
  final int height;
  final String? name;

  @override
  bool operator ==(Object other) =>
      other is AttachmentRef &&
      other.attachmentId == attachmentId &&
      other.mediaType == mediaType &&
      other.bytes == bytes &&
      other.width == width &&
      other.height == height &&
      other.name == name;

  @override
  int get hashCode =>
      Object.hash(attachmentId, mediaType, bytes, width, height, name);
}

/// One downloaded image: its durable reference plus decoded bytes.
final class AttachmentData {
  AttachmentData({required this.ref, required this.data});

  final AttachmentRef ref;
  final Uint8List data;

  @override
  bool operator ==(Object other) =>
      other is AttachmentData &&
      other.ref == ref &&
      _bytesEqual(other.data, data);

  @override
  int get hashCode => Object.hash(ref, data.length);

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}

/// Host admission limits for images, mirrored from the `imageLimits`
/// session projection. Null fields fall back to client-side defaults until
/// the host projection arrives.
final class ImageLimits {
  const ImageLimits({
    this.maxImageBytes = defaultMaxImageBytes,
    this.maxImagesPerMessage = defaultMaxImagesPerMessage,
    this.maxMessageImageBytes = defaultMaxImageBytes,
    this.maxImagePixels = defaultMaxImagePixels,
    this.mediaTypes = defaultMediaTypes,
  });

  static const int defaultMaxImageBytes = 5 * 1024 * 1024;
  static const int defaultMaxImagesPerMessage = 20;
  static const int defaultMaxImagePixels = 30000000;
  static const List<String> defaultMediaTypes = [
    'image/png',
    'image/jpeg',
    'image/webp',
    'image/gif',
  ];

  final int maxImageBytes;
  final int maxImagesPerMessage;
  final int maxMessageImageBytes;
  final int maxImagePixels;
  final List<String> mediaTypes;

  @override
  bool operator ==(Object other) =>
      other is ImageLimits &&
      other.maxImageBytes == maxImageBytes &&
      other.maxImagesPerMessage == maxImagesPerMessage &&
      other.maxMessageImageBytes == maxMessageImageBytes &&
      other.maxImagePixels == maxImagePixels &&
      _listEquals(other.mediaTypes, mediaTypes);

  @override
  int get hashCode => Object.hash(
    maxImageBytes,
    maxImagesPerMessage,
    maxMessageImageBytes,
    maxImagePixels,
    Object.hashAll(mediaTypes),
  );
}

bool _listEquals<T>(List<T> a, List<T> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}
